### PR TITLE
proposed spec for gap lines in GFA1

### PR DESCRIPTION
**Pull request by gf777**
As discussed today with @lh3, I am proposing to allow optional specification of subsequences directly in a path component, e.g. ptg000124l.1(397101:505533)+

We have devised a new tool (gfastats) that can take an agp file, for example from a scaffolding tool, and overlay the information on the original graph. When cuts are made within contigs, the segment is immutable, but we represent this event with coordinates associated to the component in the path, allowing to easily identify where cuts have occurred in the original segments.

### CODE REVIEW

**Comment by gf777** [comment]
Anyone can have a look at this? @lh3 

Thanks!

File: GFA1.md [hunk 1]
```diff
@@ -250,3 +250,39 @@ L	s12	-	s13	+	0M
 L	s11	+	s13	+	0M
 W	NA12878	1	chr1	0	11	>s11<s12>s13
 ```
+
+# `J` Jump/Gap line (since v1.2)
+
+While not a concept for pure DeBrujin or long-read assemblers, it is the case that paired end data
```
Comment (asl) [inline GFA1.md:256]: Can this be reworded as a spec somehow?

File: GFA1.md [hunk 2]
```diff
@@ -250,3 +250,39 @@ L	s12	-	s13	+	0M
...
+# `J` Jump/Gap line (since v1.2)
+
+While not a concept for pure DeBrujin or long-read assemblers, it is the case that paired end data 
+and external maps often order and orient contigs/vertices into scaffolds with intervening gaps. 
+To this end we introduce a gap edge described in J-lines that give an arbitrary or estimated gap distance between the two segment sequences.
+The gap is between the first segment at left and the second segment at right where the segments are oriented according to their sign indicators. 
+The next integer gives the arbitrary or expected distance between the first and second segment in their respective orientations. 
+Relationships in E-lines are fixed and known, where as in a G-line, the distance is an estimate and the line type is intended to allow one to define assembly scaffolds.
+
+J-line was added in GFA v1.2 and was not defined in the original GFAv1. In addition, since J-lines can contribute to paths (P-lines), these have been further specified in v1.2.
```
Comment (asl) [inline GFA1.md:263]: The information about "P" lines should not be here. The spec of P lines should be changed instead

File: GFA1.md [hunk 3]
```diff
@@ -250,3 +250,39 @@ L	s12	-	s13	+	0M
...
+Particularly, a new separator `;` is introduced as opposed to `,` to distinguish whether the path is intended to use have a gap or an overlap link between the two segments being considered.
+
+## Required fields
+
+| Column | Field        | Type      | Regexp                   | Description
+|--------|--------------|-----------|--------------------------|------------------
+| 1      | `RecordType` | Character | `G`                      | Record type
```
Comment (asl) [inline GFA1.md:270]: J

File: GFA1.md [hunk 4]
```diff
@@ -250,3 +250,39 @@ L	s12	-	s13	+	0M
...
+| 6      | `Overlap`    | String    | `[0-9]+`                 | Arbitrary or expected distance between the segments
+
+## Example
+
+```
+H	VN:Z:1.2
+S	11	ACCTT
+S	12	TCAAGG
+S	13	CTTGATT
+G	11	+	12	-	10
```
Comment (asl) [inline GFA1.md:284]: J

File: GFA1.md [hunk 5]
```diff
@@ -250,3 +250,39 @@ L	s12	-	s13	+	0M
...
+## Required fields
+
+| Column | Field        | Type      | Regexp                   | Description
+|--------|--------------|-----------|--------------------------|------------------
+| 1      | `RecordType` | Character | `G`                      | Record type
+| 2      | `From`       | String    | `[!-)+-<>-~][!-~]*`      | Name of segment
+| 3      | `FromOrient` | String    | `+\|-`                   | Orientation of From segment
+| 4      | `To`         | String    | `[!-)+-<>-~][!-~]*`      | Name of segment
+| 5      | `ToOrient`   | String    | `+\|-`                   | Orientation of `To` segment
+| 6      | `Overlap`    | String    | `[0-9]+`                 | Arbitrary or expected distance between the segments
```
Comment (asl) [inline GFA1.md:275]: We need to allow this to be `*`

**Comment by snurk** [comment]
The 'shortcut' tag description is missing

File: GFA1.md [hunk 6]
```diff
@@ -250,3 +250,39 @@ L	s12	-	s13	+	0M
...
+
+```
+H	VN:Z:1.2
+S	11	ACCTT
+S	12	TCAAGG
+S	13	CTTGATT
+G	11	+	12	-	10
+G	12	-	13	+	10
+G	11	+	13	+	3
+P	14	11+;12-;13+	*,*
```
Comment (lh3) [inline GFA1.md:287]: Would be good to have an example with both L and J lines like:
```txt
H	VN:Z:1.2
S	11	ACCTT
S	12	TCAAGG
S	13	CTTGATT
L	11	+	12	-	0M
J	12	-	13	+	10
P	14	11+,12-;13+	*
```
As to the placeholder for CIGAR, we may use dot `.` as in VCF:
```
P	14	11+,12-;13+	0M,.
```
In practice the placeholder is more for theoretical completeness. When we come to the scaffolding stage, we rarely need cigar on a P-line to disambiguate multiple edges, so most of time, we will use a single `*`.

Comment (snurk) [reply GFA1.md:287]: @lh3 but potential ambiguity is a very good point. It looks like we have to allow to specify estimated gap size in the path specification to mimic GFA1 and use this as a ‘backdoor’ to distinguish between several jumps between same nodes.

File: README.md [hunk 7]
```diff
@@ -17,6 +17,7 @@ We are developing the specification of the Graphical Fragment Assembly (GFA) for
 + [gfalint](https://github.com/sjackman/gfalint)
 + [GfaPy](https://github.com/ggonnella/gfapy)
 + [GfaViz](https://github.com/ggonnella/gfaviz)
++ [gfastats](https://github.com/vgl-hub/gfastats)
```
Comment (asl) [inline README.md:20]: Please separate out unrelated changes into a separate PR

File: GFA1.md [hunk 8]
```diff
@@ -11,14 +11,14 @@ The master version of this document can be found at
...
+The GFA format is a tab-delimited text format for describing a set of sequences and their overlap. The text is encoded in UTF-8 but is not allowed to use a codepoint value higher than 127. The first field of the line identifies the type of the line. Header lines start with `H`. Segment lines start with `S`. Link lines start with `L`. Jump lines (since v1.2) start with `J`. A containment line starts with `C`. A path line starts with `P`. Walk lines (since v1.1) start with `W`.
 
 ## Terminology
 
 + **Segment**: a continuous sequence or subsequence.
-+ **Link**: an overlap between two segments. Each link is from the end of one segment to the beginning of another segment. The link stores the orientation of each segment and the amount of basepairs overlapping.
++ **Link** or **Jump**: a connection between two oriented segments. Each connection is from the end of one oriented segment to the beginning of another oriented segment. Links store the amount of basepairs overlapping, while jumps provide estimated distance between the segments.
```
Comment (lh3) [inline GFA1.md:19]: It may be cleaner to separate Link and Jump. Also, would be good to say a main use case of Jump is for assembly gaps between contigs.

File: GFA1.md [hunk 9]
```diff
@@ -210,11 +211,44 @@ The resulting path is:
...
+If specified, the `Overlaps` field uses the `[-+]?[0-9]+J` format (note the `J` at the end to match the style of a `CIGAR` string) to refer to the jump connection with a particular estimated distance, and `.` if corresponding `J`-line does not provide distance estimate.
+
+| Column | Field          | Type      | Regexp                    | Description
+|--------|----------------|-----------|---------------------------|--------------------
+| 1      | `RecordType`   | Character | `P`                       | Record type
+| 2      | `PathName`     | String    | `[!-)+-<>-~][!-~]*`       | Path name
+| 3      | `SegmentNames` | String    | `[!-)+-<>-~][!-~]*`       | A comma/semicolon-separated list of segment names and orientations
+| 4      | `Overlaps`     | String    | `\*\|([0-9]+[MIDNSHPX=]\|\[-+]?[0-9]+J\|.)+` | Optional comma-separated list of CIGAR strings and distance estimates
+
+*TODO suggestion to allow using `.` instead of any part of `Overlaps` that we do not actually need to disambiguate*
```
Comment (lh3) [inline GFA1.md:227]: I like the idea. Do we want to include the extension in this PR? I think either way we'd better not have a "TODO" in the final spec.

File: GFA1.md [hunk 10]
```diff
@@ -250,3 +284,48 @@ L	s12	-	s13	+	0M
...
+| 3      | `FromOrient` | String    | `+\|-`                   | Orientation of From segment
+| 4      | `To`         | String    | `[!-)+-<>-~][!-~]*`      | Name of segment
+| 5      | `ToOrient`   | String    | `+\|-`                   | Orientation of `To` segment
+| 6      | `Distance`   | String    | `\*\|[-+]?[0-9]+`        | Optional estimated distance between the segments
+
+## Optional fields
+
+| Tag  | Type | Description
+|------|------|------------
+| `SC` | `Z`  | Marker tag for indirect shortcut connections. Value is ignored
```
Comment (lh3) [inline GFA1.md:323]: `Z` is encoded as a NULL-terminated string. A tiny bit less efficient. Perhaps use the `i` (integer) type here. With `i`, an option is to use non-zero for true and zero for false. Just laying out options. I am neutral.

**Comment by lh3** [comment]
CC @sjackman: the HPRC/T2T/VGP teams want to introduce a new "Jump" line type to encode assembly gaps. This also involves (backward) compatible changes to the "Path" line. If you have time, you may have a look. Thanks!

File: GFA1.md (hunk 9 above)
Comment (snurk) [reply GFA1.md]: Let's postpone.

File: GFA1.md (hunk 10 above)
Comment (snurk) [reply GFA1.md]: Changed to `i` with 0/1 values to be able to easily filter shortcuts with grep.

Comment (lh3) [reply GFA1.md]: Sound good!

File: GFA1.md (hunk 9 above)
Comment (lh3) [reply GFA1.md]: Good to me.

**Comment by snurk** [comment]
@lh3 @asl I think I addressed all the raised issues. Do you have any other comments or requests for this PR?

**Review by asl — APPROVED** [review]
LGTM, thanks!

**Comment by lh3** [comment]
Merged!

**Comment by snurk** [comment]
Thank you all!
